dataRetriever/resolvers/trieNodeResolver.go

```diff
@@ -75,28 +75,76 @@ func (tnRes *TrieNodeResolver) ProcessReceivedMessage(message p2p.MessageP2P, fr
 
 	switch rd.Type {
 	case dataRetriever.HashType:
-		var serializedNodes [][]byte
-		serializedNodes, err = tnRes.trieDataGetter.GetSerializedNodes(rd.Value, maxBuffToSendTrieNodes)
+		return tnRes.resolveOneHash(rd.Value, message)
+	case dataRetriever.HashArrayType:
+		return tnRes.resolveMultipleHashes(rd.Value, message)
+	default:
+		return dataRetriever.ErrRequestTypeNotImplemented
+	}
+}
+
+func (tnRes *TrieNodeResolver) resolveMultipleHashes(hashesBuff []byte, message p2p.MessageP2P) error {
+	b := batch.Batch{}
+	err := tnRes.marshalizer.Unmarshal(&b, hashesBuff)
+	if err != nil {
+		return err
+	}
+	hashes := b.Data
+
+	remainingSpace := maxBuffToSendTrieNodes
+	nodes := make([][]byte, 0, maxBuffToSendTrieNodes)
+	for _, hash := range hashes {
+		nextNodes, err := tnRes.getSubTrie(hash, remainingSpace)
 		if err != nil {
-			tnRes.ResolverDebugHandler().LogFailedToResolveData(
-				tnRes.topic,
-				rd.Value,
-				err,
-			)
+			continue
+		}
+
+		nodes = append(nodes, nextNodes...)
 
-			return err
+		lenNextNodes := uint64(len(nextNodes))
+		if remainingSpace >= lenNextNodes {
+			remainingSpace -= uint64(len(nextNodes))
 		}
 
-		var buff []byte
-		buff, err = tnRes.marshalizer.Marshal(&batch.Batch{Data: serializedNodes})
-		if err != nil {
-			return err
+		if lenNextNodes == 0 || remainingSpace == 0 {
+			break
```

Comment:
this is a good thing to stop until you have reached the maximum capacity to send
The condition needs to be changed. As we might have remainingSpace = 1 which is not enough to send something else. Maybe add an else branch on line 107 that will set a bool that will be used on line 109

Reply:
good catch. done